File: resources/ios/PillGroupRenderer.swift
import SwiftUI
import UIKit

enum PillGroupRendererEvent: Equatable {
    case press(callbackId: Int, nodeId: Int)

    var wireName: String { "PRESS" }
}

struct PillGroupRendererEvents {
    let send: @MainActor (PillGroupRendererEvent) -> Void

    static let native = PillGroupRendererEvents { event in
        switch event {
        case let .press(callbackId, nodeId):
            NativeUIBridge.sendPressEvent(callbackId, nodeId: nodeId)
        }
    }
}

struct PillGroupRendererConfiguration: Equatable {
    let nodeID: Int
    let optionValues: [String]
    let optionLabels: [String]
    let optionEnabled: [Bool]
    let optionCallbacks: [Int]
    let selectedValues: [String]
    let valueType: String
    let multiple: Bool
    let disabled: Bool
    let label: String
    let helper: String
    let error: String
    let required: Bool
    let accessibilityLabel: String
    let accessibilityHint: String

    init(node: NativeUINode) {
        let props = node.props
        let values = props.getStringList("option_values")

        self.nodeID = node.id
        self.optionValues = values
        self.optionLabels = props.getStringList("option_labels")
        self.optionEnabled = props.getStringList("option_enabled").map { $0 == "1" }
        self.optionCallbacks = props.getStringList("option_callbacks").map { Int($0) ?? 0 }
        self.selectedValues = props.getStringList("selected_values")
        self.valueType = props.getString("value_type")
        self.multiple = props.getBool("multiple")
        self.disabled = props.getBool("disabled") || values.isEmpty
        self.label = props.getString("label")
        self.helper = props.getString("helper")
        self.error = props.getString("error")
        self.required = props.getBool("required")
        let explicitAccessibilityLabel = props.getString("a11y_label")
        self.accessibilityLabel = explicitAccessibilityLabel.isEmpty
            ? props.getString("label")
            : explicitAccessibilityLabel
        self.accessibilityHint = props.getString("a11y_hint")
    }
}

struct PillGroupRendererState: Equatable {
    var configuration: PillGroupRendererConfiguration
    private(set) var isAwaitingPublication = false

    init(node: NativeUINode) {
        configuration = PillGroupRendererConfiguration(node: node)
    }

    mutating func userSelected(_ index: Int) -> PillGroupRendererEvent? {
        guard !isAwaitingPublication,
              !configuration.disabled,
              configuration.optionValues.indices.contains(index),
              configuration.optionEnabled.indices.contains(index),
              configuration.optionEnabled[index],
              configuration.optionCallbacks.indices.contains(index),
              configuration.optionCallbacks[index] != 0
        else {
            return nil
        }

        isAwaitingPublication = true

        return .press(
            callbackId: configuration.optionCallbacks[index],
            nodeId: configuration.nodeID
        )
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else {
            return false
        }

        let previousSelection = configuration.selectedValues
        configuration = PillGroupRendererConfiguration(node: node)
        isAwaitingPublication = false

        return previousSelection != configuration.selectedValues
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id { return node }

        for child in node.children {
            if let match = findNode(id: id, in: child) { return match }
        }

        return nil
    }
}

struct PillGroupRenderer: View {
    @ObservedObject private var themeStore = NativeUITheme.shared
    @ObservedObject private var bridge = NativeUIBridge.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var rendererState: PillGroupRendererState

    private let events: PillGroupRendererEvents

    init(node: NativeUINode, events: PillGroupRendererEvents = .native) {
        _rendererState = State(initialValue: PillGroupRendererState(node: node))
        self.events = events
    }

    var body: some View {
        let configuration = rendererState.configuration
        let theme = themeStore.resolve(for: colorScheme)

        FirstlightPillGroupField(
            label: configuration.label,
            helper: configuration.helper,
            error: configuration.error,
            required: configuration.required,
            labels: configuration.optionLabels,
            values: configuration.optionValues,
            optionEnabled: configuration.optionEnabled,
            selectedValues: configuration.selectedValues,
            disabled: configuration.disabled,
            awaitingPublication: rendererState.isAwaitingPublication,
            tokens: FirstlightPillGroupTokens.from(
                theme: theme,
                traits: UITraitCollection(
                    userInterfaceStyle: colorScheme == .dark ? .dark : .light
                )
            ),
            accessibilityLabel: configuration.accessibilityLabel,
            accessibilityHint: configuration.accessibilityHint,
            onSelection: { index in
                guard let event = rendererState.userSelected(index) else { return }
                events.send(event)
            }
        )
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            rendererState.serverPublished(tree)
        }
    }
}

File: resources/android/PillGroupRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.SelectableChipColors
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.compositeOver
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme
import com.nativephp.plugins.native_ui.NativeUITokens

sealed interface PillGroupRendererEvent {
    val wireName: String

    data class Press(
        val callbackId: Int,
        val nodeId: Int,
    ) : PillGroupRendererEvent {
        override val wireName = "PRESS"
    }
}

class PillGroupRendererEvents(
    private val send: (PillGroupRendererEvent) -> Unit,
) {
    fun dispatch(event: PillGroupRendererEvent) = send(event)

    companion object {
        val native = PillGroupRendererEvents { event ->
            when (event) {
                is PillGroupRendererEvent.Press -> NativeUIBridge.sendPressEvent(
                    event.callbackId,
                    event.nodeId,
                )
            }
        }
    }
}

data class PillGroupRendererConfiguration(
    val nodeId: Int,
    val optionValues: List<String>,
    val optionLabels: List<String>,
    val optionEnabled: List<Boolean>,
    val optionCallbacks: List<Int>,
    val selectedValues: List<String>,
    val valueType: String,
    val multiple: Boolean,
    val disabled: Boolean,
    val label: String,
    val helper: String,
    val error: String,
    val required: Boolean,
    val accessibilityLabel: String,
    val accessibilityHint: String,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        optionValues = node.props.getStringList("option_values"),
        optionLabels = node.props.getStringList("option_labels"),
        optionEnabled = node.props.getStringList("option_enabled").map { it == "1" },
        optionCallbacks = node.props.getStringList("option_callbacks").map { it.toIntOrNull() ?: 0 },
        selectedValues = node.props.getStringList("selected_values"),
        valueType = node.props.getString("value_type"),
        multiple = node.props.getBool("multiple"),
        disabled = node.props.getBool("disabled") || node.props.getStringList("option_values").isEmpty(),
        label = node.props.getString("label"),
        helper = node.props.getString("helper"),
        error = node.props.getString("error"),
        required = node.props.getBool("required"),
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty {
            node.props.getString("label")
        },
        accessibilityHint = node.props.getString("a11y_hint"),
    )
}

class PillGroupRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(PillGroupRendererConfiguration(node))
        private set

    var isAwaitingPublication by mutableStateOf(false)
        private set

    fun userSelected(index: Int): PillGroupRendererEvent? {
        if (isAwaitingPublication ||
            configuration.disabled ||
            !configuration.optionValues.indices.contains(index) ||
            !configuration.optionEnabled.getOrElse(index) { false }
        ) {
            return null
        }

        val callback = configuration.optionCallbacks.getOrNull(index)?.takeIf { it != 0 }
            ?: return null

        isAwaitingPublication = true
        return PillGroupRendererEvent.Press(callback, configuration.nodeId)
    }

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findPillGroupNode(configuration.nodeId) ?: return false
        val previousSelection = configuration.selectedValues
        configuration = PillGroupRendererConfiguration(node)
        isAwaitingPublication = false
        return previousSelection != configuration.selectedValues
    }
}

object PillGroupRenderer {
    @Composable
    fun Render(
        node: NativeUINode,
        modifier: Modifier,
        events: PillGroupRendererEvents = PillGroupRendererEvents.native,
    ) {
        val rendererState = remember(node.id) { PillGroupRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value

        LaunchedEffect(publishedTree) {
            publishedTree?.let(rendererState::serverPublished)
        }

        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light

        FirstlightPillGroupControl(
            labels = rendererState.configuration.optionLabels,
            values = rendererState.configuration.optionValues,
            enabled = rendererState.configuration.optionEnabled,
            selectedValues = rendererState.configuration.selectedValues,
            groupEnabled = !rendererState.configuration.disabled,
            awaitingPublication = rendererState.isAwaitingPublication,
            label = rendererState.configuration.label,
            helper = rendererState.configuration.helper,
            error = rendererState.configuration.error.takeIf(String::isNotEmpty),
            onSelection = { index ->
                rendererState.userSelected(index)?.let(events::dispatch)
            },
            modifier = modifier,
            colors = pillGroupColors(tokens),
            labelColor = tokens.onSurface,
            helperColor = tokens.onSurfaceVariant,
            errorColor = tokens.destructive,
            required = rendererState.configuration.required,
            accessibilityLabel = rendererState.configuration.accessibilityLabel,
            accessibilityHint = rendererState.configuration.accessibilityHint,
        )
    }
}

private fun NativeUINode.findPillGroupNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findPillGroupNode(id)?.let { return it } }
    return null
}

internal data class PillGroupTokenColors(
    val selectedContainer: Color,
    val selectedContent: Color,
    val unselectedContainer: Color,
    val unselectedContent: Color,
    val outline: Color,
)

internal fun resolvePillGroupTokenColors(tokens: NativeUITokens): PillGroupTokenColors {
    val surface = tokens.surface.compositeOver(tokens.background).copy(alpha = 1f)
    val selectedContainer = tokens.primary.compositeOver(surface).copy(alpha = 1f)
    val preferredSelectedContent = tokens.onPrimary.compositeOver(selectedContainer).copy(alpha = 1f)
    val selectedContent = preferredSelectedContent.takeIf {
        contrastRatio(it, selectedContainer) >= 4.5f
    } ?: listOf(Color.Black, Color.White).maxBy { contrastRatio(it, selectedContainer) }

    return PillGroupTokenColors(
        selectedContainer = selectedContainer,
        selectedContent = selectedContent,
        unselectedContainer = tokens.surfaceVariant.compositeOver(surface).copy(alpha = 1f),
        unselectedContent = tokens.onSurface,
        outline = tokens.outline,
    )
}

@Composable
internal fun pillGroupColors(tokens: NativeUITokens): SelectableChipColors {
    val resolved = resolvePillGroupTokenColors(tokens)

    return FilterChipDefaults.filterChipColors(
        containerColor = resolved.unselectedContainer,
        labelColor = resolved.unselectedContent,
        selectedContainerColor = resolved.selectedContainer,
        selectedLabelColor = resolved.selectedContent,
        disabledContainerColor = resolved.unselectedContainer.copy(alpha = 0.38f),
        disabledLabelColor = tokens.onSurfaceVariant.copy(alpha = 0.38f),
    )
}
